implementation/alpha/node/node.ino

```diff
@@ -322,58 +332,47 @@ void loop() {
     if (my_data.has_sensor && (do_first_reading_packet || millis() - last_reading_time >= PACKET_PERIOD)) {
-        do_first_reading_packet = false;
-
-        Packet new_packet = {
-            .reading = {},
-            .age = 0,
-            .number = packet_number,
-            .origin_id = my_id,
-            .current_id = my_id,
-        };
-
-        read_temperatures(new_packet.reading);
-
-        // Clear space by deleting older packets
-        if (packet_queue.size == QUEUE_SIZE_MAX) {
-            packet_queue.pop();
-        }
-
-        packet_queue.push(new_packet);
-
-        if (PRINT_DEBUG) {
-            Serial.print("Packet created: ");
-            Serial.println(packet_number);
-            print_packet(new_packet);
-        }
-
-        packet_number++;
-
-        last_reading_time = millis();
-    } else if (my_id > 128) {
-        health_packet_generate();
-    }
-
-    // Set frequency
-    float expected_frequency = 0.0f;
-    if (packet_queue.size > 0 && my_data.parent != NO_ID) {
-        expected_frequency = parent_data.rx_frequency;
-    } else {
-        expected_frequency = my_data.rx_frequency;
-    }
-    if (current_frequency != expected_frequency) {
-        rf69.setFrequency(expected_frequency);
-        current_frequency = expected_frequency;
-    }
-
-    if (packet_queue.size > 0 && my_data.parent != NO_ID) {
-        loop_tx();
-    } else {
-        loop_rx();
-    }
-
-    blink_led_periodically();
-
-    // Update packet ages
-    updatePacketAge(millis() - loop_start_time);
+    data_packet_generate();
+  } else if (my_id > 128 &&
+             millis() - last_healthPacket_time >= HEALTH_PACKET_PERIOD) {
+    health_packet_generate();
+  }
+  // Set frequency
+  float expected_frequency = 0.0f;
+  if (nextState == Transmit && packet_queue.size > 0 &&
+      my_data.parent != NO_ID) {
+    expected_frequency = parent_data.rx_frequency;
+  } else if (nextState == Receive && !my_data.has_sensor) {
+    expected_frequency = my_data.rx_frequency;
+  }
+  if (current_frequency != expected_frequency) {
+    rf69.setFrequency(expected_frequency);
+    current_frequency = expected_frequency;
+  }
+  if (packet_queue.size > 0 && my_data.parent != NO_ID) {
+    loop_tx();
+  } else {
+    loop_rx();
+  }
+  // For debug purposes
+  blink_led_periodically();
+  // Update packet ages
+  updatePacketAge(millis() - loop_start_time);
+  // Unified sleep
+  if (millis() - last_transmit >= current_tx_sleep) {
+    current_tx_sleep = convert_to_sleepydog_time(expovariate(TX_RATE));
+    next_transmit = millis() + current_tx_sleep;
+  }
+
+  if (millis() - last_receive >= current_rx_sleep) {
+    current_rx_sleep = convert_to_sleepydog_time(expovariate(RX_RATE));
+    next_receive = millis() + current_rx_sleep;
+  }
+  if (!my_data.has_sensor && next_transmit - millis() > next_receive - millis()) {
+    nextState = Receive;
+    delay(next_receive - millis());
+  } else {
+    nextState = Transmit;
```

Comment:
loop() is always delaying/sleeping for TX. If the node has a sensor but doesn't have any packets to send, then last_transmit will not be updated(loop_tx is not run) and  next_transmit will be less than millis(). This results in a negative delay value, or in uint32_t terms, an extremely large value.